### PR TITLE
Endian modes

### PR DESCRIPTION
**Pull request by KayEss**
Have a native mode together with the explicit LE and BE ones.

### CODE REVIEW

File: include/felspar/parse/extract.detail.hpp [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+#include <span>
+
+
+namespace felspar::parse::binary::detail {
+
+
+    template<typename T>
+    inline T native_extract(std::span<std::byte const, sizeof(T)> const s) noexcept {
+        T into;
+        std::copy(s.begin(), s.end(), reinterpret_cast<std::byte *>(&into));
```
Comment (jtv) [inline include/felspar/parse/extract.detail.hpp]: Is that _legal?_  It doesn't need any kind of memory barrier to tell the compiler that there's deliberate aliasing of incompatible types?  Would `std::as_writable_bytes` help with that?

Also, how portable do you need the communication to be?  C++ now guarantees 2's complement for integers... but I don't think it guarantees much about floating-point formats.

File: include/felspar/parse/extract.native.hpp [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+    inline T extract(std::span<std::byte const> &s,
+                    felspar::source_location const &loc =
+                            felspar::source_location::current()) {
+        if (s.size() >= sizeof(T)) {
+            auto const v = unchecked_extract<T>(std::span<std::byte const, sizeof(T)>{s.data(), sizeof(T)});
+            s = s.subspan(sizeof(T));
+            return v;
+        } else {
+            throw felspar::stdexcept::logic_error{
+                    "Buffer doesn't have enough data", loc};
```
Comment (jtv) [inline include/felspar/parse/extract.native.hpp]: Might be helpful to give a bit more context for debugging here: what were you trying to extract, how many bytes did you need, how many bytes did you get?

(There's no guaranteed way to determine a human-readable name for a type, but gcc and clang environments have `__cxa_demangle`, and Visual Studio likes human-readable mangling.  And if we're talking about a very limited set of types, there's always template variable specialisation.)

File: include/felspar/parse/insert.detail.hpp [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+#include <span>
+
+
+namespace felspar::parse::binary::detail {
+
+
+    template<typename T>
+    inline void native_insert(
+            std::span<std::byte, sizeof(T)> const s, T const t) noexcept {
+        auto const addr = reinterpret_cast<std::byte const *>(&t);
```
Comment (jtv) [inline include/felspar/parse/insert.detail.hpp]: Again, aren't you supposed to tell the compiler when you're type-punning?  Would `std::as_bytes` help?

File: include/felspar/parse/extract.detail.hpp (hunk 1 above)
Comment (KayEss) [reply include/felspar/parse/extract.detail.hpp]: There's likely some weird platforms where floats aren't IEEE, but I don't think that'll be the case for any that I care about. The stand out is Windows where apparently `long double` isn't 128 bits. That may make that type non-portable, although that's not really anything to do with this library.

https://en.cppreference.com/w/cpp/language/types

File: include/felspar/parse/extract.native.hpp (hunk 2 above)
Comment (KayEss) [reply include/felspar/parse/extract.native.hpp]: In practice it's not very useful as the result is generally that the buffer has zero bytes in it and any error happened much earlier, but I've added the sizes and abstracted out the actual size check as well.

File: include/felspar/parse/insert.detail.hpp (hunk 3 above)
Comment (KayEss) [reply include/felspar/parse/insert.detail.hpp]: Yeah, I've changed these now and made sure everything is consistent.

**Review by jtv — APPROVED** [review]
Thanks for making the changes.

**Pull request merged**